cupy/core/core.pyx

```diff
@@ -1862,9 +1862,10 @@ cdef int _cuda_runtime_version = -1
 cdef str _cuda_path = ''  # '' for uninitialized, None for non-existing
 
 cdef list _cupy_header_list = [
-    'cupy/complex.cuh',
-    'cupy/carray.cuh',
     'cupy/atomics.cuh',
+    'cupy/carray.cuh',
```

Comment:
`carray.cuh` has to be first, because it defines the `float16` class which is used in `atomics.cuh`. A simple fix would be to revert the order, but we can also consider splitting fp16 support to a separate header.